Libraries/LibWeb/CredentialManagement/CredentialsContainer.cpp

```diff
@@ -17,25 +20,476 @@ GC::Ref<CredentialsContainer> CredentialsContainer::create(JS::Realm& realm)
+
+    // 8. For each interface of interfaces:
+    for (auto& interface : interfaces) {
+        // 1. If options.mediation is conditional and interface does not support conditional user mediation,
+        //    return a promise rejected with a "TypeError" DOMException.
+        if (options.mediation == Bindings::CredentialMediationRequirement::Conditional && !interface->supports_conditional_user_mediation())
+            return WebIDL::create_rejected_promise(realm(), JS::TypeError::create(realm(), "Conditional user mediation is not supported"sv));
+
+        // 2. If settings’ active credential types contains interface’s [[type]],
+        //    return a promise rejected with a "NotAllowedError" DOMException.
+        if (settings.active_credential_types().contains_slow(interface->type()))
+            return WebIDL::create_rejected_promise_from_exception(realm(), WebIDL::NotAllowedError::create(realm(), "Credential type is not allowed"_string));
+
+        // 3. Append interface’s [[type]] to settings’ active credential types.
+        settings.active_credential_types().append(interface->type());
+    }
+
+    // 9. Let origin be settings’ origin.
+    auto origin = settings.origin();
+
+    // 10. Let sameOriginWithAncestors be true if settings is same-origin with its ancestors, and false otherwise.
+    auto same_origin_with_ancestors = is_same_origin_with_its_ancestors(settings);
+
+    // 11. For each interface in options’ relevant credential interface objects:
+    for (auto& interface : interfaces) {
+        // 1. Let permission be the interface’s [[type]] Get Permissions Policy.
+        auto permission = interface->get_permission_policy();
+
+        // 2. If permission is null, continue.
+        if (!permission.has_value())
+            continue;
+
+        // TODO: 3. If document is not allowed to use permission, return a promise rejected with a "NotAllowedError" DOMException.
+    }
+
+    // 12. Let p be a new promise.
+    auto promise = WebIDL::create_promise(realm());
+
+    // 13. Run the following steps in parallel:
+    Platform::EventLoopPlugin::the().deferred_invoke(GC::create_function(realm().heap(), [this, promise = GC::Root(promise), origin, options = move(options), same_origin_with_ancestors, &settings] {
+        HTML::TemporaryExecutionContext execution_context { realm(), HTML::TemporaryExecutionContext::CallbacksEnabled::Yes };
+
+        // 1. Let credentials be the result of collecting Credentials from the credential store, given origin, options, and sameOriginWithAncestors.
+        auto maybe_credentials = collect_credentials_from_store(realm(), origin, options, same_origin_with_ancestors);
+
+        // 2. If credentials is an exception, reject p with credentials.
+        if (maybe_credentials.is_error()) {
+            WebIDL::reject_promise(realm(), *promise, maybe_credentials.error_value());
```

Comment:
There is a specification bug here that is pretty far-reaching. Any time we are "in parallel", it is invalid to interact with JS. This promise rejection must be wrapped in a queued task.

https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-for-spec-authors

There's a w3c bot that has filed issues about this in other repos. I asked on whatwg matrix if they'll run it against Credential Management.

We can come back and fix this later :tm: once the spec is fixed. since it doesnt' actually have an effect in our implementation (I think).

Reply:
I'll put some FIXMEs for that.